### ISSUE TITLE
[RFC] : Prediction Resolver Core Refactor

### ISSUE THREAD

**Issue by Khanz9664** [Khanz9664/TrustLens#89]
## Motivation

TrustLens currently relies on implicit scikit-learn-centric assumptions around prediction APIs (`predict`, `predict_proba`) and probability shape semantics.

While effective for sklearn estimators, this coupling limits extensibility to broader ML ecosystems (XGBoost, Keras, TensorFlow, PyTorch) and increases the risk of framework-specific logic leaking into core trustworthiness evaluation.

To support the planned roadmap, TrustLens must decouple prediction resolution from trust evaluation.

This refactor introduces a framework-agnostic architecture that standardizes model outputs into a unified contract before analysis begins.

---

## Proposed Architecture
The refactor introduces three distinct tiers:
1. **Orchestration Tier (`api.py`)**: High-level public entry point.
2. **Standardization Tier (`trustlens.backends`)**: A functional registry that standardizes model outputs (tensors, objects) into `PredictionBundle` objects (numpy arrays).
3. **Engine Tier (`trustlens.core.pipeline`)**: A framework-agnostic pipeline that computes metrics purely from standardized data.

### Architecture Flow

```text
User
  │
  ▼
analyze() ──────────────────────┐
(api.py)                        │
  │                             │
  ▼                             │
resolve_predictions()           │
(backends/registry.py)          │
  │                             │
  ▼                             │
PredictionBundle                │
(y_pred, y_prob, metadata)      │
  │                             │
  ▼                             │
_run_analysis_pipeline()        │
(core/pipeline.py)              │
  │                             │
  ├── Calibration               │
  ├── Failure Analysis          │
  ├── Fairness/Bias             │
  ├── Representation            │
  └── Plugins                   │
          │
          ▼
      TrustReport
```

### PR Roadmap
We will execute this refactor in 8 atomic phases to ensure safe delivery and bit-for-bit parity:

- **PR 0: Characterization Tests** - Baseline existing behavior.
- **PR 1: Core Pipeline Extraction** - Move metric logic to `core/pipeline.py` (Behavioral parity).
- **PR 2: Standardization Contract** - Implement `types.py` and the `Prediction Semantics Contract`.
- **PR 3: Sklearn Resolver** - Implement the first functional backend.
- **PR 4: Registry & Detection Layer** - Implement module-name-based framework detection.
- **PR 5: Integration (Wiring Only)** - Connect `api.py` to the new flow.
- **PR 6: Metadata & Auditability** - Add framework and version tracking to `TrustReport`.
- **PR 7: First Extension (XGBoost)** - Demonstrate extensibility.

---

### Prediction Semantics Contract
All backends must adhere to the rules defined in `docs/internal/prediction_contract.md`:
- **Logits Policy**: Raw logits are strictly forbidden; backends must normalize to probabilities (Softmax/Sigmoid).
- **Shape Normalization**: Binary classification must be normalized to `(n, 2)`.
- **Labels**: `class_labels` must be preserved to ensure report semantics.

### Framework Detection
We utilize a deterministic detection priority:
1. Explicit `framework` argument.
2. Module-name inspection (`type(model).__module__`).
3. Capability detection (`hasattr`).

---

## XAI Compatibility

The `core/pipeline.py` engine may receive references to the `model` and `X` to support future explainability workflows (e.g., GradCAM, saliency, counterfactual methods).

However, strict architectural isolation is enforced:

> `core/pipeline.py` must never invoke framework-specific model methods (`predict`, `predict_proba`, `forward`, etc.).

The pipeline may pass model references to downstream explainability modules, but remains framework-agnostic itself.

This preserves decoupling while enabling future deep learning and XAI extensions.

---

## Expected Outcomes
- **Zero breaking changes** for existing scikit-learn users.
- **Clean contributor path** for new framework support (XGBoost, Keras, etc.).
- **Enhanced auditability** with framework metadata in reports.
- **Robustness** against future technical debt through strict interface boundaries.

---

## Non-Goals

This RFC does **not** introduce:

- XGBoost support (implemented in PR 7)
- TensorFlow/Keras support
- PyTorch support
- Regression task support
- Deep learning explainability features
- Changes to TrustLens public APIs
- Breaking behavior changes for existing sklearn users

The objective is architectural decoupling only.

---

## Success Criteria

This refactor is considered complete when:

- Existing sklearn workflows produce identical outputs (bit-for-bit parity).
- `analyze()` behavior remains unchanged for existing users.
- Framework-specific prediction logic is isolated to `trustlens.backends`.
- `_run_analysis_pipeline()` becomes framework-agnostic.
- `import trustlens` does not introduce optional ML dependencies.
- XGBoost support can be added without modifying core analysis logic.

---

### Why not use `isinstance(model, ...)` for framework detection?

`isinstance()` introduces optional dependency coupling.

Example:

```python
import tensorflow as tf

isinstance(model, tf.keras.Model)
```

This requires TensorFlow to be installed and imported, increasing import overhead and coupling optional dependencies to core TrustLens execution.

Instead, TrustLens uses:

1. Explicit `framework=` override
2. Module-name inspection
3. Capability detection

This preserves lazy dependency loading and improves extensibility.

**Comment by saeidsaadatigero** [comment]
Hi @Khanz9664,

This is an exceptionally well-structured RFC. Decoupling the prediction resolution layer from the evaluation engine using a functional registry and standardizing into a `PredictionBundle` is definitely the right architectural move to avoid framework leakage.

As a Senior Python and AI Engineer with deep experience in designing scalable ML pipelines, clean OOP architectures, and working extensively with scikit-learn/PyTorch estimators, I would love to contribute to this refactor.

Given the atomic roadmap you provided, I can start right away with either **PR 0 (Characterization Tests)** to ensure bit-for-bit parity, or jump straight into **PR 2 / PR 3 (Standardization Contract & Sklearn Resolver)**. 

Could you please assign this issue (or the first couple of phases) to me? Looking forward to collaborating!

**Comment by Khanz9664** [comment]
@saeidsaadatigero Thanks a lot for the thoughtful feedback — I really appreciate the detailed read of the RFC.

Glad the architecture direction resonates with your experience. This refactor is fairly foundational for TrustLens, so I’m intentionally executing the early phases in a tightly controlled sequence to preserve behavioral parity and architectural consistency.

PR0 (Characterization Tests) has already been completed and I’m currently moving into PR1 (Core Pipeline Extraction). For the first few phases (PR1–PR5), I’ll likely keep implementation centralized since they establish the core abstractions (`PredictionBundle`, registry design, pipeline isolation, framework detection, etc.) that future backends will depend on.

That said, I’d definitely love collaboration as the architecture stabilizes. A great place to jump in would be:

* reviewing the prediction contract and resolver interface as it evolves
* helping validate parity and edge cases
* contributing to later modular phases (e.g., metadata/auditability, XGBoost backend support, or future framework integrations)

I’ll keep this issue updated as milestones land and would be happy to coordinate on a scoped contribution once the core layer is in place.

**Comment by saeidsaadatigero** [comment]
Hi @Khanz9664,

Sounds like a solid plan! It completely makes sense to keep the foundation (PR1-PR5) centralized to ensure architectural integrity. 

I would be more than happy to help review the prediction contract as it evolves and jump into the modular phases later on—especially the XGBoost backend support and validating edge cases, as that aligns perfectly with my AI/ML background.

I'll keep an eye on this issue for updates. Good luck with the core implementation!

**Comment by Khanz9664** [comment]
Thanks @saeidsaadatigero — really appreciate the thoughtful follow-up and the willingness to contribute.

The core refactor is progressing well so far (currently through the sklearn backend phase), and I’m intentionally keeping the foundational architecture tightly scoped until the registry/integration layers stabilize.

Once the core prediction resolver flow is fully in place, XGBoost support and backend validation will be a great place to collaborate. Edge-case testing around prediction semantics, probability normalization, and framework-specific quirks will be especially valuable.

I’ll continue updating the RFC issue as milestones land — looking forward to collaborating once we reach the extension phase.
